pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
+			if err != nil {
+				return nil, err
+			}
+
+			if len(pv) > 0 {
+				patch[key] = pv
+			}
+		case string, float64, bool:
+			if !matchesValue(av, bv) {
+				patch[key] = bv
+			}
+		case []interface{}:
+			bt := bv.([]interface{})
+			if fieldPatchStrategy == "merge" {
+				pv, err := diffLists(at, bt, fieldType.Elem(), fieldPatchMergeKey)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(pv) > 0 {
+					patch[key] = pv
+				}
+			} else if !matchesArray(at, bt, t) {
+				patch[key] = bv
+			}
+		case nil:
+			switch bv.(type) {
+			case nil:
+				// Both nil, fine.
+			default:
+				patch[key] = bv
+			}
+		default:
+			panic(fmt.Sprintf("Unknown type:%T in key %s", av, key))
+		}
+	}
+
+	// Now add all deleted values as nil
+	for key := range a {
+		_, found := b[key]
+		if !found {
+			patch[key] = nil
+		}
+	}
+
+	return patch, nil
+}
+
+// Returns true if the array matches (must be json types).
+// As is idiomatic for go, an empty array is not the same as a nil array.
+func matchesArray(a, b []interface{}, t reflect.Type) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	if (a == nil && b != nil) || (a != nil && b == nil) {
+		return false
+	}
+
+	for i := range a {
+		if !matchesValue(a[i], b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Returns true if the values matches (must be json types)
+// The types of the values must match, otherwise it will always return false
+// If two map[string]interface{} are given, all elements must match.
+func matchesValue(av, bv interface{}) bool {
+	switch at := av.(type) {
+	case string:
+		bt, ok := bv.(string)
+		if ok && (bt == at) {
+			return true
+		}
+	case float64:
+		bt, ok := bv.(float64)
+		if ok && (bt == at) {
+			return true
+		}
+	case bool:
+		bt, ok := bv.(bool)
+		if ok && (bt == at) {
+			return true
+		}
+	case map[string]interface{}:
+		bt, ok := bv.(map[string]interface{})
+		if ok && (len(at) == len(bt)) {
+			for key := range at {
+				if !matchesValue(at[key], bt[key]) {
+					return false
+				}
+			}
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// Returns a (recursive) strategic merge patch that yields b when applied to a,
+// for a pair of lists with merge semantics.
+func diffLists(a, b []interface{}, t reflect.Type, mergeKey string) ([]interface{}, error) {
+	if len(a) == 0 {
+		if len(b) == 0 {
+			return nil, nil
+		}
+
+		return b, nil
+	}
+
+	et, err := sliceElementType(a, b)
+	if err != nil {
+		return nil, err
+	}
+
+	var patch []interface{}
+
+	// If the elements are not maps...
+	if et.Kind() == reflect.Map {
+		patch, err = diffListsOfMaps(a, b, t, mergeKey)
+	} else {
+		patch, err = diffListsOfScalars(a, b)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// Returns a (recursive) strategic merge patch that yields b when applied to a,
+// for a pair of lists of scalars with merge semantics.
+func diffListsOfScalars(a, b []interface{}) ([]interface{}, error) {
```

Comment:
Actually, a list of scalars should just be completely replaced in a patch, since normally lists in JSON are replaced all at once. Only if it is a merging list do we actually have to generate a patch, and we don't support merging lists of scalars so we never need to diff lists of scalars. In both cases above, the patch should simply return the new list value, e.g. for 1., the patch should return `[1, 2, 3]` and for 2. the patch should return `[1, 2, 2, 2, 3]`.